File: prototype/rails/src/app/domain/cart/cart_quantity.rb
module Domain
  module Cart
    module CartQuantity
      module_function

      # A cart never holds more of a listing than the seller has left.
      def within_stock(requested:, available:)
        raise ArgumentError, "a cart holds at least one of a listing, got #{requested}" if requested < 1
        raise ArgumentError, "that listing is sold out" if available < 1

        [requested, available].min
      end
    end
  end
end

File: prototype/node/src/app/core/cart/cart-quantity.ts
/**
 * A cart never holds more of a listing than the seller has left. `isPurchasable`
 * settles sold out at the gate in front of the add, inside the transaction that
 * writes the line, so a throw here marks a caller that added past the gate.
 */
export function quantityWithinStock(input: { requested: number; available: number }): number {
  const { requested, available } = input

  if (requested < 1) {
    throw new RangeError(`a cart holds at least one of a listing, got ${requested}`)
  }
  if (available < 1) {
    throw new RangeError('that listing is sold out')
  }

  return Math.min(requested, available)
}
